field/TinyLang/Field/Typed/Core.hs

```diff
@@ -228,32 +257,55 @@ isTracked uniq x env =
         Nothing -> False
 
 -- TODO: test me somehow.
-exprVarSigns :: Expr f a -> ScopedVarSigns f
-exprVarSigns = goExpr $ ScopedVarSigns mempty mempty where
-    goStat :: ScopedVarSigns f -> Statement f -> ScopedVarSigns f
-    goStat signs (ELet uniVar def) = ScopedVarSigns free $ insertUnique uniq sign bound where
-        UniVar uni (Var uniq name) = uniVar
-        sign = VarSign name uni
-        ScopedVarSigns free bound = goExpr signs def
-    goStat signs (EAssert expr) = goExpr signs expr
-
-    goExpr :: ScopedVarSigns f -> Expr f a -> ScopedVarSigns f
-    goExpr signs (EConst _) = signs
-    goExpr signs (EVar (UniVar uni (Var uniq name)))
-        | tracked   = signs
-        | otherwise = ScopedVarSigns (insertUnique uniq sign free) bound
-        where
-            ScopedVarSigns free bound = signs
-            sign = VarSign name uni
-            tracked = isTracked uniq sign bound || isTracked uniq sign free
-    goExpr signs (EAppUnOp _ x) = goExpr signs x
-    goExpr signs (EAppBinOp _ x y) = goExpr (goExpr signs x) y
-    goExpr signs (EIf b x y) = goExpr (goExpr (goExpr signs b) x) y
-    goExpr signs (EStatement st e) = goExpr (goStat signs st) e
-
-exprFreeVarSigns :: Expr f a -> Env (VarSign f)
-exprFreeVarSigns = _scopedVarSignsFree . exprVarSigns
-
-supplyFromAtLeastFree :: MonadSupply m => Expr f a -> m ()
-supplyFromAtLeastFree =
-    supplyFromAtLeast . freeUniqueIntMap . unEnv . _scopedVarSignsFree . exprVarSigns
+stmtVarSigs' :: ScopedVarSigs f -> Statement f -> ScopedVarSigs f
+stmtVarSigs' sigs (ELet uniVar def) = ScopedVarSigs free $ insertUnique uniq sig bound where
+    UniVar uni (Var uniq name) = uniVar
+    sig = VarSig name uni
+    ScopedVarSigs free bound = exprVarSigs' sigs def
+stmtVarSigs' sigs (EAssert expr) = exprVarSigs' sigs expr
+stmtVarSigs' sigs (EFor uniVar _ _ stmts) = ScopedVarSigs free $ insertUnique uniq sig bound where
+    UniVar uni (Var uniq name) = uniVar
+    sig = VarSig name uni
+    ScopedVarSigs free bound = foldr (flip stmtVarSigs') sigs (C.unStatements stmts)
```

Comment:
Here's the bug, it seems. `foldr` processes statements in reverse order, so a bound variable may become added to the set of free variables. If you look at this line below:

```haskell
exprVarSigs' sigs (EIf b x y) = exprVarSigs' (exprVarSigs' (exprVarSigs' sigs b) x) y
```

then it becomes clear that we have a left fold here, so `foldl` should be used here.

Reply:
Ah, I remember looking at that line, and the issue with sticking of `Expr f a` in a fold, is that `a` might be different, therefore it will not be possible to easily stuck them in a list, so I am not sure they can be stuck in a foldable.

Reply:
Yeah, I'm not suggesting doing that, just giving an example.

Reply:
Fixed now.